db/db_impl.cc

```diff
@@ -2296,15 +2296,17 @@ Status DBImpl::FlushMemTable(ColumnFamilyData* cfd,
       return Status::OK();
     }
 
-    WriteThread::Writer w;
-    write_thread_.EnterUnbatched(&w, &mutex_);
+    if (!cfd->mem()->IsEmpty()) {
+      WriteThread::Writer w;
+      write_thread_.EnterUnbatched(&w, &mutex_);
 
-    // SwitchMemtable() will release and reacquire mutex
-    // during execution
-    s = SwitchMemtable(cfd, &context);
-    write_thread_.ExitUnbatched(&w);
+      // SwitchMemtable() will release and reacquire mutex
+      // during execution
+      s = SwitchMemtable(cfd, &context);
+      write_thread_.ExitUnbatched(&w);
 
-    cfd->imm()->FlushRequested();
+      cfd->imm()->FlushRequested();
+    }
 
     // schedule flush
```

Comment:
check out line 2316. you will still wait for the memtable flush to finish, even though you never scheduled it (since memtable was empty). that's a bit tricky situation. we should not wait for flush if we don't need it. I see four different cases:
1. memtable is full -- switch the memtable, request the flush, etc.
2. memtable empty, immutable memtable non-empty, some immutable memtables are not getting flushed -- request the flush of immutable memtables, wait until it finishes
3. memtable empty, immutable memtable non-empty, all immutable memtables are getting flushed -- don't request the flush, but i think we'll still need to wait for the immutable memtable flush to finish
4. memtable empty, immutable memtables empty -- don't do anythign